whylog/log_reader/__init__.py

```diff
@@ -47,25 +47,37 @@ def _save_clues_in_normal_dict(cls, collector):
             (parser_name, list(clues_iter)) for parser_name, clues_iter in collector.items()
         )
 
+    def _constraints_verification(self, clues):
+        """
+        provides constraints verification basing on rules from investigation_plan
+        and collected clues
+        """
+        causes = []
+        # effect_clues are accessible from self._investigation_plan._effect_clues
+        for rule in self._investigation_plan._suspected_rules:
+            clues_lists = []
+            for cause in rule._causes:
+                if cause.name in clues:
+                    clues_lists.append(clues[cause.name])
+            effect_clue = self._investigation_plan._effect_clues[rule._effect.name]
+            results_list = Verifier.constraints_and(clues_lists, effect_clue, rule._constraints)
```

Comment:
Look you have calls someting like rule._causes, rule._effect_name. rule._constraints. I think that lines [58-63] should be method in rule class. You can name it check_constraints or something like that. After this refactor ugly calls disapear